snorkel/mtl/model.py

```diff
@@ -0,0 +1,402 @@
+import logging
+import os
+from collections import defaultdict
+from collections.abc import Iterable
+
+import numpy as np
+import torch
+import torch.nn as nn
+
+from snorkel.mtl.task import Task
+from snorkel.mtl.utils import move_to_device, prob_to_pred, recursive_merge_dicts
+
+model_default_config = {
+    "model_path": None,  # the path to a saved checkpoint to initialize with
+    "device": 0,  # gpu id (int) or -1 for cpu
+    "dataparallel": True,
+}
+
+
+class MultitaskModel(nn.Module):
+    """A class to build multi-task model.
+
+    :param name: Name of the model
+    :type name: str
+    :param tasks: a list of Task that trains jointly
+    :type tasks: list of Task project
+    """
+
+    def __init__(self, name=None, tasks=None, **kwargs):
+        super().__init__()
+        self.config = recursive_merge_dicts(
+            model_default_config, kwargs, misses="insert"
+        )
+        self.name = name if name is not None else type(self).__name__
+
+        # Initiate the model attributes
+        self.module_pool = nn.ModuleDict()
+        self.task_names = set()
+        self.task_flows = dict()
```

Comment:
How often are task flows used in full generality, versus following simple templates?  Am assuming this is part of what you meant when you said this could be simplified?  To discuss later!

Reply:
We actually did use some of the fancier logic for the SuperGLUE tasks (e.g., when you have a multiple choice question and you need to run all four options through the network separately first before making a prediction). I think the interface definitely could use some work though, and we should be able to make special simplified syntax for the most common cases at the very least.